src/neighborsearch/bounded.jl

```diff
@@ -3,39 +3,122 @@
 # ------------------------------------------------------------------
 
 """
-    BoundedSearch(method, nmax)
+    BoundedSearch(method, maxneighbors=Inf; maxpercategory,
+    maxperoctant, ordermetric=Euclidean())
 
-A method for searching at most `nmax` neighbors using `method`.
+A method for searching at most `maxneighbors` neighbors using `method`. Extra
+restrictions available: `maxpercategory` and `maxperoctant`. The priority is
+given to the nearest neighbor using `ordermetric`.
 """
 struct BoundedSearch{M<:NeighborSearchMethod} <: BoundedNeighborSearchMethod
   method::M
-  nmax::Int
+  maxneighbors::Int
+  maxpercategory::Dict{Symbol,Int}
+  maxperoctant::Int
+  ordermetric::Metric
+
+  function BoundedSearch{M}(method::M, maxneighbors=0; maxpercategory=Dict(),
+                            maxperoctant=0, ordermetric=Euclidean()) where {M}
+    new(method, maxneighbors, maxpercategory, maxperoctant, ordermetric)
+  end
 end
 
+BoundedSearch(method::M, maxneighbors::Int=0; kwargs...) where {M} =
+  BoundedSearch{M}(method, maxneighbors; kwargs...)
+
 object(method::BoundedSearch) = object(method.method)
 
-maxneighbors(method::BoundedSearch) = method.nmax
+maxneighbors(method::BoundedSearch) = method.maxneighbors
 
 function search!(neighbors, xₒ::AbstractVector,
                  method::BoundedSearch; mask=nothing)
-  inds = search(xₒ, method.method)
-  nmax = method.nmax
+  # get reference objects and initial neighbors
+  meth = method.method
+  obj  = object(method)
+  N    = ncoords(obj)
+  inds = search(xₒ, meth, mask=mask)
+
+  # get distances and give priority to closest neighbors according to ordermetric
+  ometric = method.ordermetric
+  dists   = colwise(ometric, xₒ, coordinates(obj,inds))
+  sortids = sortperm(dists)
+  inds    = inds[sortids]
+
+  # read constraints
+  maxk   = method.maxneighbors
+  maxoct = method.maxperoctant
+  maxcat = method.maxpercategory
+  usek   = maxk   > 0
+  useoct = maxoct > 0
+  usecat = length(maxcat) > 0
```

Comment:
If these variables are only used once in the conditional statements, maybe they shouldn't exist. For example, in the next section below we could just write

```julia
# handle octant constraint
if maxoct > 0`
  ...
end
```

Reply:
Oh I see that some of these are used more than once. If you feel that it would be nicer to organize them in a section like this, that is fine too. Just make sure that the section is highlighted and not mixed with other parts of the function body.